src/webserver/DOSGuard.h

```diff
@@ -9,11 +9,48 @@
 class DOSGuard
 {
     boost::asio::io_context& ctx_;
-    std::mutex mtx_;  // protects ipFetchCount_
+    std::recursive_mutex
+        mtx_;  // protects ipFetchCount_, ipRequestCount_
+               // recursive mutex allows reentrant mutex acquisitions
+               // this makes isOk(ip) and checkout(ip) atomic
     std::unordered_map<std::string, std::uint32_t> ipFetchCount_;
+    std::unordered_map<std::string, std::uint32_t> ipRequestCount_;
     std::unordered_set<std::string> const whitelist_;
     std::uint32_t const maxFetches_;
     std::uint32_t const sweepInterval_;
+    std::uint32_t const maxConcurrentRequests_;
+
+    struct RPCScope
+    {
+        std::uint32_t& count_;
+        std::recursive_mutex& mtx_;
+        bool isValid_;
+
+        RPCScope(std::uint32_t& ct, std::recursive_mutex& m, bool isValid)
+            : count_(ct), mtx_(m), isValid_(isValid)
+        {
+            if (isValid_)
+            {
+                std::unique_lock lk(mtx_);
+                count_++;
+            }
+        }
+
+        ~RPCScope()
+        {
+            if (isValid_)
```

Comment:
Should we `count_` an RPC towards the `count_` even if it is invalid? 

Up to you, functionally I don't think it changes a lot.